src/wallet/walletdb.cpp

```diff
@@ -297,426 +298,573 @@ bool WalletBatch::EraseLockedUTXO(const COutPoint& output)
+    }
 
-                    // Extract the index and internal from the path
-                    // Path string is m/0'/k'/i'
-                    // Path vector is [0', k', i'] (but as ints OR'd with the hardened bit
-                    // k == 0 for external, 1 for internal. i is the index
-                    if (path.size() != 3) {
-                        strErr = "Error reading wallet database: keymeta found with unexpected path";
-                        return false;
-                    }
-                    if (path[0] != 0x80000000) {
-                        strErr = strprintf("Unexpected path index of 0x%08x (expected 0x80000000) for the element at index 0", path[0]);
-                        return false;
-                    }
-                    if (path[1] != 0x80000000 && path[1] != (1 | 0x80000000)) {
-                        strErr = strprintf("Unexpected path index of 0x%08x (expected 0x80000000 or 0x80000001) for the element at index 1", path[1]);
-                        return false;
-                    }
-                    if ((path[2] & 0x80000000) == 0) {
-                        strErr = strprintf("Unexpected path index of 0x%08x (expected to be greater than or equal to 0x80000000)", path[2]);
-                        return false;
+    // Load unencrypted keys
+    LoadResult key_res = LoadRecords(pwallet, batch, DBKeys::KEY,
+        [] (CWallet* pwallet, DataStream& key, CDataStream& value, std::string& err) {
+        return LoadKey(pwallet, key, value, err) ? DBErrors::LOAD_OK : DBErrors::CORRUPT;
+    });
+    result = std::max(result, key_res.m_result);
+
+    // Load encrypted keys
+    LoadResult ckey_res = LoadRecords(pwallet, batch, DBKeys::CRYPTED_KEY,
+        [] (CWallet* pwallet, DataStream& key, CDataStream& value, std::string& err) {
+        return LoadCryptedKey(pwallet, key, value, err) ? DBErrors::LOAD_OK : DBErrors::CORRUPT;
+    });
+    result = std::max(result, ckey_res.m_result);
+
+    // Load scripts
+    LoadResult script_res = LoadRecords(pwallet, batch, DBKeys::CSCRIPT,
+        [] (CWallet* pwallet, DataStream& key, CDataStream& value, std::string& strErr) {
+        uint160 hash;
+        key >> hash;
+        CScript script;
+        value >> script;
+        if (!pwallet->GetOrCreateLegacyScriptPubKeyMan()->LoadCScript(script))
+        {
+            strErr = "Error reading wallet database: LegacyScriptPubKeyMan::LoadCScript failed";
+            return DBErrors::NONCRITICAL_ERROR;
+        }
+        return DBErrors::LOAD_OK;
+    });
+    if (script_res.m_result != DBErrors::LOAD_OK) {
+        return script_res.m_result;
+    }
+    result = std::max(result, script_res.m_result);
+
+    // Check whether rewrite is needed
+    if (ckey_res.m_records > 0) {
+        // Rewrite encrypted wallets of versions 0.4.0 and 0.5.0rc:
+        if (last_client == 40000 || last_client == 50000) return DBErrors::NEED_REWRITE;
```

Comment:
I think the general principle was to try to detect as much corruption as possible before returning the error. But it would also make this a lot simpler to just abort on the first error.

Reply:
In commit "walletdb: Refactor legacy wallet record loading into its own function" (9d58900d8a754721964e628f0c44344652f4bb80)

> I think the general principle was to try to detect as much corruption as possible before returning the error. But it would also make this a lot simpler to just abort on the first error.

This is still giving NEED_REWRITE error higher priority over other errors besides CORRUPT. And it is adding an early return in the middle of a 211 line function that mostly accumulates errors instead of returning early. If this error needs to be handled differently than other errors and can't be merged with std::max, I think there should a comment calling out the special case and early return. Otherwise better to make this more consistent and write this this as:

```diff
--- a/src/wallet/walletdb.cpp
+++ b/src/wallet/walletdb.cpp
@@ -865,9 +865,9 @@ static DBErrors LoadLegacyWalletRecords(CWallet* pwallet, DatabaseBatch& batch,
     result = std::max(result, script_res.m_result);
 
     // Check whether rewrite is needed
-    if (ckey_res.m_records > 0 && result != DBErrors::CORRUPT) {
+    if (ckey_res.m_records > 0) {
         // Rewrite encrypted wallets of versions 0.4.0 and 0.5.0rc:
-        if (last_client == 40000 || last_client == 50000) return DBErrors::NEED_REWRITE;
+        if (last_client == 40000 || last_client == 50000) result = std::max(result, DBErrors::NEED_REWRITE);
     }
 
     // Load keymeta

```